File: src/main/java/com/redlimerl/detailab/api/render/BarRenderManager.java
package com.redlimerl.detailab.api.render;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.util.Identifier;
import net.minecraft.util.dynamic.Codecs;
import org.jetbrains.annotations.NotNull;

import java.awt.*;

public interface BarRenderManager {

    @NotNull
    Texture getTextureFull();

    @NotNull
    Texture getTextureHalf();

    @NotNull
    Texture getTextureOutline();

    @NotNull
    Texture getTextureOutlineHalf();

    boolean isHidden();

    @NotNull
    Color getColor();

    record Texture(Identifier location, int width, int height, TextureOffset offset) {

        public static final Codec<Texture> CODEC = RecordCodecBuilder.create(instance -> instance.group(
                Identifier.CODEC.fieldOf("texture").forGetter(Texture::location),
                Codecs.POSITIVE_INT.fieldOf("texture_width").forGetter(Texture::width),
                Codecs.POSITIVE_INT.fieldOf("texture_height").forGetter(Texture::height),
                TextureOffset.CODEC.fieldOf("offset").forGetter(Texture::offset)
        ).apply(instance, Texture::new));

    }
}

File: src/main/kotlin/com/redlimerl/detailab/api/BarRenderManager.kt
package com.redlimerl.detailab.api

import net.minecraft.util.Identifier
import java.awt.Color


class TextureOffset(val x: Int, val y: Int)

interface BarRenderManager {
    val texture: Identifier
    val textureWidth: Int
    val textureHeight: Int
    val textureOffsetFull: TextureOffset
    val textureOffsetHalf: TextureOffset
    val textureOffsetOutline: TextureOffset
    val textureOffsetOutlineHalf: TextureOffset
    val isShown: Boolean
    val color: Color
}

/**
 * Render options for Armor Bar.
 * @param texture Texture Identifier where the Armor Bar
 * @param textureWidth Texture's width
 * @param textureHeight Texture's height
 * @param textureOffsetFull Full Armor Bar texture position
 * @param textureOffsetHalf Half Armor Bar texture position
 * @param textureOffsetOutline Full Armor Bar Outline texture position
 * @param textureOffsetOutlineHalf Half Armor Bar Outline texture position
 * @param color If you want, change the color of the texture. But outline is not change. Default is [Color.WHITE]
 * @see TextureOffset
 */
class ArmorBarRenderManager(
    override val texture: Identifier, override val textureWidth: Int, override val textureHeight: Int,
    override val textureOffsetFull: TextureOffset, override val textureOffsetHalf: TextureOffset,
    override val textureOffsetOutline: TextureOffset, override val textureOffsetOutlineHalf: TextureOffset, override val color: Color = Color.WHITE
) : BarRenderManager {

    override val isShown: Boolean
        get() = true

}


/**
 * Render options for Wearable Item Armor Bar.
 * @param texture Texture Identifier where the Armor Bar
 * @param textureWidth Texture's width
 * @param textureHeight Texture's height
 * @param textureOffsetFull Full Armor Bar texture position
 * @param textureOffsetOutline Full Armor Bar Outline texture position
 * @param isShown Whether the Item Bar is visible.
 * @param color If you want, change the color of the texture. But outline is not change. Default is [Color.WHITE]
 * @see TextureOffset
 */
class ItemBarRenderManager(
    override val texture: Identifier, override val textureWidth: Int, override val textureHeight: Int,
    override val textureOffsetFull: TextureOffset, override val textureOffsetOutline: TextureOffset,
    override val isShown: Boolean, override val color: Color = Color.WHITE
) : BarRenderManager {

    override val textureOffsetHalf: TextureOffset
        get() = textureOffsetFull
    override val textureOffsetOutlineHalf: TextureOffset
        get() = textureOffsetOutline

}